File: ios/ArveyGo/ArveyGo/ViewModels/DashboardViewModel.swift
import SwiftUI
import Combine

// MARK: - Dashboard ViewModel
@MainActor
class DashboardViewModel: ObservableObject {
    @Published var vehicles: [Vehicle] = []
    @Published var drivers: [DriverScore] = []
    @Published var alerts: [AlarmEvent] = []
    @Published var vehiclesErrorMessage: String?
    @Published var alertsErrorMessage: String?
    @Published var selectedPeriod: String = "today"
    @Published var isLoading = true
    @Published var isRefreshing = false
    @Published var isLoadingDrivers = false
    @Published var isLoadingDailyKm = true
    @Published var isLoadingAlerts = false

    private var cancellables = Set<AnyCancellable>()
    private let wsManager = WebSocketManager.shared

    var totalVehicles: Int { vehicles.count }
    /// Kontak açık: ignition == true (online + idle)
    var kontakOnCount: Int { vehicles.filter { $0.ignition }.count }
    /// Kontak kapalı: isOnline ama ignition == false
    var kontakOffCount: Int { vehicles.filter { $0.isOnline && !$0.ignition }.count }
    /// Bilgi yok: isOnline == false (cihazdan veri gelmiyor)
    var bilgiYokCount: Int { vehicles.filter { !$0.isOnline }.count }
    var onlineCount: Int { vehicles.filter { $0.status == .ignitionOn }.count }
    var offlineCount: Int { vehicles.filter { $0.status == .ignitionOff || $0.status == .noData }.count }
    var idleCount: Int { vehicles.filter { $0.status == .sleeping }.count }
    var totalKm: Int { vehicles.reduce(0) { $0 + $1.totalKm } }
    var todayKm: Int { vehicles.reduce(0) { $0 + $1.todayKm } }
    var avgScore: Int {
        guard !drivers.isEmpty else { return 0 }
        return drivers.reduce(0) { $0 + $1.score } / drivers.count
    }

    var metrics: [DashboardMetric] {
        [
            DashboardMetric(
                title: "Toplam Araç",
                value: "\(totalVehicles)",
                icon: "car.fill",
                iconBg: AppTheme.navy.opacity(0.06),
                iconColor: AppTheme.navy,
                change: "Değişim yok",
                changeType: .flat
            ),
            DashboardMetric(
                title: "Kontak Açık",
                value: "\(kontakOnCount)",
                icon: "key.fill",
                iconBg: AppTheme.online.opacity(0.08),
                iconColor: AppTheme.online,
                change: "",
                changeType: .flat
            ),
            DashboardMetric(
                title: "Kontak Kapalı",
                value: "\(kontakOffCount)",
                icon: "key",
                iconBg: AppTheme.idle.opacity(0.08),
                iconColor: AppTheme.idle,
                change: "",
                changeType: .flat
            ),
            DashboardMetric(
                title: "Bilgi Yok",
                value: "\(bilgiYokCount)",
                icon: "questionmark.circle.fill",
                iconBg: AppTheme.offline.opacity(0.08),
                iconColor: AppTheme.offline,
                change: "",
                changeType: .flat
            ),
            DashboardMetric(
                title: "Bugün Km",
                value: formatKm(todayKm),
                icon: "road.lanes",
                iconBg: AppTheme.indigo.opacity(0.08),
                iconColor: AppTheme.indigo,
                change: "",
                changeType: .flat
            ),
        ]
    }

    init() {
        subscribeToWebSocket()
        loadVehiclesFromAPI()
        loadDriversFromAPI()
        loadAlertsFromAPI()
    }

    // MARK: - WebSocket Subscription
    private func subscribeToWebSocket() {
        // Observe vehicle list from WebSocketManager
        wsManager.$vehicleList
            .receive(on: DispatchQueue.main)
            .sink { [weak self] list in
                guard let self = self else { return }
                if !list.isEmpty {
                    // Mevcut araç değerlerini koruyarak güncelle (null sıcaklık/nem için)
                    let currentMap = Dictionary(uniqueKeysWithValues: self.vehicles.map { ($0.id, $0) })
                    self.vehicles = list.map { newVehicle in
                        if var existing = currentMap[newVehicle.id] {
                            existing.mergeUpdate(from: newVehicle)
                            return existing
                        }
                        return newVehicle
                    }
                    self.vehiclesErrorMessage = nil
                    self.isLoading = false
                    self.isLoadingDailyKm = false
                }
            }
            .store(in: &cancellables)

        // Also listen for individual events
        wsManager.eventSubject
            .receive(on: DispatchQueue.main)
            .sink { [weak self] event in
                guard let self = self else { return }
                switch event {
                case .snapshot(let vehicles, _, _):
                    let currentMap = Dictionary(uniqueKeysWithValues: self.vehicles.map { ($0.id, $0) })
                    self.vehicles = vehicles.map { newVehicle in
                        if var existing = currentMap[newVehicle.id] {
                            existing.mergeUpdate(from: newVehicle)
                            return existing
                        }
                        return newVehicle
                    }
                    self.vehiclesErrorMessage = nil
                    self.isLoading = false
                    self.isLoadingDailyKm = false
                case .update(let vehicle, _):
                    if let idx = self.vehicles.firstIndex(where: { $0.id == vehicle.id }) {
                        self.vehicles[idx].mergeUpdate(from: vehicle)
                    } else {
                        self.vehicles.append(vehicle)
                    }
                default: break
                }
            }
            .store(in: &cancellables)
    }

    func formatKm(_ km: Int) -> String {
        let formatter = NumberFormatter()
        formatter.numberStyle = .decimal
        formatter.groupingSeparator = "."
        return formatter.string(from: NSNumber(value: km)) ?? "\(km)"
    }

    /// Fetch vehicle list from REST API and merge dailyKm data
    func loadVehiclesFromAPI() {
        isLoadingDailyKm = true
        vehiclesErrorMessage = nil
        Task {
            do {
                let apiVehicles = try await APIService.shared.fetchVehicles()
                // Merge: prefer WS live positions, but take dailyKm from API
                let currentMap = Dictionary(uniqueKeysWithValues: vehicles.map { ($0.id, $0) })
                let merged = apiVehicles.map { apiV -> Vehicle in
                    if var existing = currentMap[apiV.id] {
                        existing.mergeUpdate(from: apiV)
                        if apiV.dailyKm > existing.dailyKm { existing.dailyKm = apiV.dailyKm }
                        if apiV.todayKm > existing.todayKm { existing.todayKm = apiV.todayKm }
                        return existing
                    }
                    return apiV
                }
                vehicles = merged
                vehiclesErrorMessage = nil
                isLoading = false
                isLoadingDailyKm = false
            } catch {
                print("[DashboardVM] fetchVehicles error: \(error)")
                vehiclesErrorMessage = error.localizedDescription
                isLoading = false
                isLoadingDailyKm = false
            }
        }
    }

    /// Load drivers from API and convert to DriverScore for dashboard display
    func loadDriversFromAPI() {
        isLoadingDrivers = true
        Task {
            do {
                let response = try await APIService.shared.fetchDrivers()
                let avatarColors: [SwiftUI.Color] = [
                    AppTheme.navy, AppTheme.indigo, AppTheme.online, .blue,
                    AppTheme.idle, AppTheme.lavender, AppTheme.offline, .gray,
                    .purple, .orange, .teal, .pink
                ]
                let scores: [DriverScore] = response.drivers
                    .sorted { $0.scoreGeneral > $1.scoreGeneral }
                    .enumerated()
                    .map { (index, driver) in
                        DriverScore(
                            id: driver.id,
                            name: driver.name,
                            plate: driver.vehicle.isEmpty ? driver.lastVehicle : driver.vehicle,
                            score: driver.scoreGeneral,
                            totalKm: Int(driver.totalDistanceKm),
                            color: avatarColors[index % avatarColors.count]
                        )
                    }
                await MainActor.run {
                    self.drivers = scores
                    self.isLoadingDrivers = false
                }
            } catch {
                print("[DashboardVM] fetchDrivers error: \(error)")
                await MainActor.run {
                    self.isLoadingDrivers = false
                }
            }
        }
    }

    /// Load alerts from API
    func loadAlertsFromAPI() {
        isLoadingAlerts = true
        alertsErrorMessage = nil
        Task {
            do {
                let json = try await APIService.shared.get("/api/mobile/alarms?per_page=5")
                let dataArr = json["data"] as? [[String: Any]] ?? []
                let alertList = dataArr.enumerated().map { index, item in
                    AlarmEvent.from(json: item, index: index)
                }
                await MainActor.run {
                    self.alerts = alertList
                    self.alertsErrorMessage = nil
                    self.isLoadingAlerts = false
                }
            } catch {
                print("[DashboardVM] fetchAlarms error: \(error)")
                await MainActor.run {
                    self.alerts = []
                    self.alertsErrorMessage = "Alarm verisi şu anda alınamıyor."
                    self.isLoadingAlerts = false
                }
            }
        }
    }

    func setPeriod(_ period: String) {
        selectedPeriod = period
        // In real app, this would fetch data for the period
    }

    /// Pull-to-refresh: WS'yi yeniden bağla ve verileri yenile
    func refreshData() {
        isRefreshing = true
        wsManager.reconnect()
        loadVehiclesFromAPI()
        loadDriversFromAPI()
        loadAlertsFromAPI()
        // 2 saniye sonra refreshing durumunu kapat
        DispatchQueue.main.asyncAfter(deadline: .now() + 2) { [weak self] in
            self?.isRefreshing = false
        }
    }
}

File: android/app/src/main/java/com/arveya/arveygo/viewmodels/DashboardViewModel.kt
package com.arveya.arveygo.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.arveya.arveygo.models.*
import com.arveya.arveygo.services.APIService
import com.arveya.arveygo.services.WebSocketManager
import com.arveya.arveygo.ui.theme.AppColors
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch
import java.text.NumberFormat
import java.util.Locale

class DashboardViewModel : ViewModel() {
    private val _vehicles = MutableStateFlow<List<Vehicle>>(emptyList())
    val vehicles: StateFlow<List<Vehicle>> = _vehicles

    private val _drivers = MutableStateFlow<List<DriverScore>>(emptyList())
    val drivers: StateFlow<List<DriverScore>> = _drivers

    private val _alerts = MutableStateFlow<List<AlarmEvent>>(emptyList())
    val alerts: StateFlow<List<AlarmEvent>> = _alerts
    private val _vehiclesErrorMessage = MutableStateFlow<String?>(null)
    val vehiclesErrorMessage: StateFlow<String?> = _vehiclesErrorMessage
    private val _alertsErrorMessage = MutableStateFlow<String?>(null)
    val alertsErrorMessage: StateFlow<String?> = _alertsErrorMessage

    private val _isLoadingDrivers = MutableStateFlow(false)
    val isLoadingDrivers: StateFlow<Boolean> = _isLoadingDrivers

    private val _isLoadingDailyKm = MutableStateFlow(true)
    val isLoadingDailyKm: StateFlow<Boolean> = _isLoadingDailyKm

    private val _isLoadingAlerts = MutableStateFlow(false)
    val isLoadingAlerts: StateFlow<Boolean> = _isLoadingAlerts

    private val _isLoading = MutableStateFlow(true)
    val isLoading: StateFlow<Boolean> = _isLoading

    private val _selectedPeriod = MutableStateFlow("today")
    val selectedPeriod: StateFlow<String> = _selectedPeriod

    val totalVehicles: Int get() = _vehicles.value.size
    val onlineCount: Int get() = _vehicles.value.count { it.status == VehicleStatus.IGNITION_ON }
    val offlineCount: Int get() = _vehicles.value.count { it.status == VehicleStatus.IGNITION_OFF || it.status == VehicleStatus.NO_DATA }
    val idleCount: Int get() = _vehicles.value.count { it.status == VehicleStatus.SLEEPING }
    val kontakOnCount: Int get() = _vehicles.value.count { it.ignition }
    val kontakOffCount: Int get() = _vehicles.value.count { it.isOnline && !it.ignition }
    val bilgiYokCount: Int get() = _vehicles.value.count { !it.isOnline }
    val totalKm: Int get() = _vehicles.value.sumOf { it.totalKm }
    val todayKm: Int get() = _vehicles.value.sumOf { it.todayKm }
    val avgScore: Int get() {
        val d = _drivers.value
        return if (d.isEmpty()) 0 else d.sumOf { it.score } / d.size
    }

    private val _isRefreshing = MutableStateFlow(false)
    val isRefreshing: StateFlow<Boolean> = _isRefreshing

    fun refreshData() {
        viewModelScope.launch {
            _isRefreshing.value = true
            WebSocketManager.reconnect()
            loadVehiclesFromAPI()
            loadDriversFromAPI()
            loadAlertsFromAPI()
            delay(2000)
            _isRefreshing.value = false
        }
    }

    fun getMetrics(): List<DashboardMetric> = listOf(
        DashboardMetric("Toplam Araç", "$totalVehicles", "car", AppColors.Navy.copy(alpha = 0.06f), AppColors.Navy, "", ChangeType.FLAT),
        DashboardMetric("Kontak Açık", "$kontakOnCount", "check_circle", AppColors.Online.copy(alpha = 0.08f), AppColors.Online, "", ChangeType.FLAT),
        DashboardMetric("Kontak Kapalı", "$kontakOffCount", "cancel", AppColors.Idle.copy(alpha = 0.08f), AppColors.Idle, "", ChangeType.FLAT),
        DashboardMetric("Bilgi Yok", "$bilgiYokCount", "pause_circle", AppColors.Offline.copy(alpha = 0.08f), AppColors.Offline, "", ChangeType.FLAT),
        DashboardMetric("Bugün Km", formatKm(todayKm), "road", AppColors.Indigo.copy(alpha = 0.08f), AppColors.Indigo, "", ChangeType.FLAT),
    )

    fun formatKm(km: Int): String = NumberFormat.getNumberInstance(Locale("tr", "TR")).format(km)

    fun setPeriod(period: String) { _selectedPeriod.value = period }

    init {
        subscribeToWebSocket()
        loadVehiclesFromAPI()
        loadDriversFromAPI()
        loadAlertsFromAPI()
    }

    private fun subscribeToWebSocket() {
        // Single source of truth: observe vehicle list from WebSocketManager
        viewModelScope.launch {
            WebSocketManager.vehicleList.collectLatest { list ->
                if (list.isNotEmpty()) {
                    _vehicles.value = list
                    _vehiclesErrorMessage.value = null
                    _isLoading.value = false
                    _isLoadingDailyKm.value = false
                }
            }
        }
    }

    private fun loadVehiclesFromAPI() {
        viewModelScope.launch {
            _isLoadingDailyKm.value = true
            _vehiclesErrorMessage.value = null
            try {
                val apiVehicles = APIService.fetchVehicles()
                // Merge API data into existing vehicle list, preserving WS live data
                val wsMap = _vehicles.value.associateBy { it.id }
                val merged = apiVehicles.map { apiV ->
                    val wsV = wsMap[apiV.id]
                    if (wsV != null) {
                        wsV.copy(
                            todayKm = if (apiV.todayKm > wsV.todayKm) apiV.todayKm else wsV.todayKm,
                            dailyKm = if (apiV.dailyKm > wsV.dailyKm) apiV.dailyKm else wsV.dailyKm,
                            lastPacketAt = apiV.lastPacketAt ?: wsV.lastPacketAt,
                            deviceTime = apiV.deviceTime ?: wsV.deviceTime
                        )
                    } else apiV
                }
                _vehicles.value = merged
                _vehiclesErrorMessage.value = null
                _isLoading.value = false
            } catch (e: Exception) {
                android.util.Log.e("DashboardVM", "fetchVehicles error", e)
                _vehiclesErrorMessage.value = e.localizedMessage ?: "Araç verileri alınamadı."
                _isLoading.value = false
            }
            _isLoadingDailyKm.value = false
        }
    }

    private fun loadDriversFromAPI() {
        viewModelScope.launch {
            _isLoadingDrivers.value = true
            try {
                val response = APIService.fetchDrivers()
                val avatarColors = listOf(
                    AppColors.Navy, AppColors.Indigo, AppColors.Online, androidx.compose.ui.graphics.Color.Blue,
                    AppColors.Idle, AppColors.Lavender, AppColors.Offline, androidx.compose.ui.graphics.Color.Gray,
                    androidx.compose.ui.graphics.Color(0xFF9C27B0), androidx.compose.ui.graphics.Color(0xFFFF9800),
                    androidx.compose.ui.graphics.Color(0xFF009688), androidx.compose.ui.graphics.Color(0xFFE91E63)
                )
                _drivers.value = response.drivers
                    .sortedByDescending { it.scoreGeneral }
                    .mapIndexed { index, driver ->
                        DriverScore(
                            id = driver.id,
                            name = driver.name,
                            plate = driver.vehicle.ifEmpty { driver.lastVehicle },
                            score = driver.scoreGeneral,
                            totalKm = driver.totalDistanceKm.toInt(),
                            color = avatarColors[index % avatarColors.size]
                        )
                    }
            } catch (e: Exception) {
                android.util.Log.e("DashboardVM", "fetchDrivers error", e)
            }
            _isLoadingDrivers.value = false
        }
    }

    private fun loadAlertsFromAPI() {
        viewModelScope.launch {
            _isLoadingAlerts.value = true
            try {
                val json = APIService.get("/api/mobile/alarms?per_page=5")
                val dataArr = json.optJSONArray("data")
                if (dataArr != null && dataArr.length() > 0) {
                    val alertList = mutableListOf<AlarmEvent>()
                    for (i in 0 until dataArr.length()) {
                        alertList.add(AlarmEvent.from(dataArr.getJSONObject(i), i))
                    }
                    _alerts.value = alertList
                    _alertsErrorMessage.value = null
                } else {
                    _alerts.value = emptyList()
                    _alertsErrorMessage.value = null
                }
            } catch (e: Exception) {
                android.util.Log.e("DashboardVM", "fetchAlarms error", e)
                _alerts.value = emptyList()
                _alertsErrorMessage.value = "Alarm verileri şu anda alınamıyor."
            }
            _isLoadingAlerts.value = false
        }
    }
}
